applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/types.h>
+#include <zephyr/bluetooth/conn.h>
+
+#include <caf/events/ble_common_event.h>
+#include <caf/events/ble_smp_event.h>
+#include "config_event.h"
+#include "usb_event.h"
+#include <caf/events/button_event.h>
+#include <caf/events/power_event.h>
+
+#define MODULE dvfs
+#include <caf/events/module_state_event.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_DVFS_LOG_LEVEL);
+
+#include "ld_dvfs_handler.h"
+
+#define DVFS_INIT_TIMEOUT		K_SECONDS(1)
+#define REG_CONN_INTERVAL_LLPM_MASK	0x0d00
+
+enum dvfs_states_list {
+	DVFS_STATE_DEFAULT,
+	DVFS_STATE_LLPM,
+	DVFS_STATE_USB,
+	DVFS_STATE_COUNT
+};
+
+/* Table containing the frequency settings for each state.
+ *It is tailored for nrf54h20dk and might need to be adjusted for other platforms.
+ */
+enum dvfs_frequency_setting dvfs_freq_table[DVFS_STATE_COUNT] = {
+	DVFS_FREQ_LOW,		/* DVFS_STATE_DEFAULT */
+	DVFS_FREQ_MEDLOW,	/* DVFS_STATE_LLPM */
+	DVFS_FREQ_HIGH		/* DVFS_STATE_USB */
+	};
+
+/* Binary mask tracking which states are requested */
+static uint8_t dfvs_requests_state;
+static enum dvfs_frequency_setting current_freq = DVFS_FREQ_HIGH;
+/* DVFS inits after a while, so we need to wait for it to be ready */
+static struct k_work_delayable DVFS_init_timeout;
+
+static const char *get_dvfs_frequency_setting_name(enum dvfs_frequency_setting setting)
+{
+	switch (setting) {
+	case DVFS_FREQ_HIGH: return "DVFS_FREQ_HIGH";
+	case DVFS_FREQ_MEDLOW: return "DVFS_FREQ_MEDLOW";
+	case DVFS_FREQ_LOW: return "DVFS_FREQ_LOW";
+	default: return "Unknown";
+	}
+}
+
+static void set_dvfs_freq(enum dvfs_frequency_setting target_freq)
+{
+	size_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+
+	if (ret == -EAGAIN) {
+		LOG_INF("DVFS not initialized, try again.");
+	} else if (ret == -EBUSY) {
+		LOG_INF("DVFS frequency change in progress");
```

Comment:
We should retry on error after a timeout (otherwise we may get stuck with invalid frequency)